ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,341 @@
+
+let files_in dir ~otherwise =
+  Lwt.catch
+    (fun () ->
+      let* listing = Lwt_unix.files_of_directory dir |> Lwt_stream.to_list in
+      List.filter_map
+        (function "." | ".." -> None | name -> Some (dir // name))
+        listing
+      |> Lwt.return
+    )
+    otherwise
+
+let unlink_safe file =
+  let __FUN = __FUNCTION__ in
+  Lwt.catch
+    (fun () -> Lwt_unix.unlink file)
+    (function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          Lwt.pause ()
+      | e ->
+          Logs_lwt.info (fun m ->
+              m "%s: error %s when deleting %s, ignoring" __FUN
+                (Printexc.to_string e) file
+          )
+      )
+
+(** - Direct: request doesn't pass through the cache
+    - Engaged: both side coordinate through the queue, writer ends the mode
+      when the queue has been filled.
+    - Disengaged: writer ignores the queue, reader empties it and the cache;
+      then it changes the mode to engaged.
+*)
+type state = Direct | Engaged | Disengaged
+
+type channel = {
+    queue: t Lwt_bounded_stream.t
+  ; push: t option -> unit option
+  ; lock: Lwt_mutex.t (* lock for the states *)
+  ; mutable state: state
+}
+
+(*
+   Notes:
+   - uses Mtime.t to force usage of monotonic time
+   - This means that between runs (and reboots) cached stated is lost if not
+     persisted first.
+   IDEA: carryover: read contents of cache and "convert it" to the current run
+
+   TODO:
+     - Add startup step to convert existing content to new time
+     - Exponential backoff on xapi push error
+     - Limit error logging on xapi push error: once per downtime is enough
+ *)
+
+module Writer : sig
+  val with_cache :
+       direct:(t -> string -> (unit, exn) Lwt_result.t)
+    -> Types.Service.t
+    -> channel
+    -> ((t -> string -> unit Lwt.t) -> 'a Lwt.t)
+    -> 'a Lwt.t
+  (** [with_cache ~direct typ context] creates a cache for content of type [typ]. The
+      cache is writable through the function [context], which is provided a
+      writing function to persist to the cache.
+
+      Example:
+        Xapi_guard.Disk_cache.(Writer.with_cache Tpm (upload session_cache)
+        @@ fun write_tpm -> write_tpm (uuid, time) contents
+    *)
+end = struct
+  let mkdir_safe name perm =
+    Lwt.catch
+      (fun () -> Lwt_unix.mkdir name perm)
+      (function
+        | Unix.(Unix_error (EEXIST, _, _)) -> Lwt.return_unit | e -> raise e
+        )
+
+  let files_in_existing dir =
+    let create_dir = function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          let* () = mkdir_safe dir 0o700 in
+          Lwt.return []
+      | e ->
+          raise e
+    in
+    files_in dir ~otherwise:create_dir
+
+  let persist_to ~filename ~contents =
+    let flags = Unix.[O_WRONLY; O_CREAT; O_SYNC] in
+    let perm = 0o600 in
+    let persist out_chan = Lwt_io.write out_chan contents in
+    Lwt_io.with_file ~flags ~perm ~mode:Output filename persist
```

Comment:
Unfortunately we don't have an equivalent for it in Lwt, but we have a Unixext function to atomically write something to a file. We might need to use that here to survive host crashes, like with HA, a regular write to file may be completely lost as it'll be done purely in-memory in the OS's page cache.

I think the sequence is:
* you write the data to a file (like you've done here)
* you call fsync
* you rename the file
* you fsync the parent directory (if you skip this step you may lose both the old and the new file as POSIX allows that behaviour...)

Reply:
Then the documentation is downright dangerous, it characterizes writes as ["atomic"](https://ocaml.org/p/lwt/latest/doc/Lwt_io/index.html#writing)

Reply:
maybe "atomic" with respect to the Lwt scheduler...

Reply:
they are atomic in the sense that writes are either submitted to the file or not, they are not atomic vs to what happens on disk (that is up to the OS)